### PR TITLE
Broadcastables

### PR DESCRIPTION
**Pull request by raver119**
***WIP; DO NOT MERGE;***

This PR adds:
- _bp for broadcastable operations
- new op tileToShape
- fixes to matmul, strided_slice ops
- new ops spaceToBatch/batchToSpace
- more permute/reshape equality tests vs ND4j

### CODE REVIEW

File: include/ops/declarable/generic/broadcastable/add.cpp [hunk 1]
```diff
@@ -61,5 +61,66 @@ namespace nd4j {
...
+                gradY->assign(tmp);
+                gradX->assign(epsNext);
+            } else {
+                // broadcast case
+                auto axisX = ShapeUtils<T>::evalBroadcastBackwardAxis(x->shapeInfo(), epsNext->shapeInfo());
+                auto axisY = ShapeUtils<T>::evalBroadcastBackwardAxis(y->shapeInfo(), epsNext->shapeInfo());
+
+                if (axisX.size() > 0) {
+                    auto sum = epsNext->template reduceAlongDimension<simdOps::Sum<T>>(axisX);
+                    gradX->assign(sum);
```
Comment (AlexDBlack) [inline include/ops/declarable/generic/broadcastable/add.cpp]: Minor thing, not sure if it's possible here:
In ND4J, we can do sum ops and specify the result array. If we can do that here, it'd save us an assign op?

File: include/ops/declarable/generic/broadcastable/floordiv.cpp [hunk 2]
```diff
@@ -86,5 +86,41 @@ namespace nd4j {
 
             return shapeList;
         }
+
+        CUSTOM_OP_IMPL(floordiv_bp, 3, 2, false, 0, 0) {
+            // PLEASE NOTE: we're just passing eps down the line here
```
Comment (AlexDBlack) [inline include/ops/declarable/generic/broadcastable/floordiv.cpp]: Old comment?

File: include/ops/declarable/generic/broadcastable/floormod.cpp [hunk 3]
```diff
@@ -86,5 +86,41 @@ namespace nd4j {
 
             return shapeList;
         }
+
+        CUSTOM_OP_IMPL(floormod_bp, 3, 2, false, 0, 0) {
+            // PLEASE NOTE: we're just passing eps down the line here
```
Comment (AlexDBlack) [inline include/ops/declarable/generic/broadcastable/floormod.cpp]: As per above.

File: include/ops/declarable/generic/broadcastable/mod.cpp [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+                auto tZ = x->template applyTrueBroadcast<simdOps::Divide<T>>(y);
+                OVERWRITE_RESULT(tZ);
+            } else {
+                auto sx = ShapeUtils<T>::shapeAsString(*x);
+                auto sy = ShapeUtils<T>::shapeAsString(*y);
+                REQUIRE_TRUE(false, 0, "Divide: shapes should be equal, or broadcastable. But got %s vs %s instead", sx.c_str(), sy.c_str());
+            }
+            */
+
+			return ND4J_STATUS_OK;
```
Comment (AlexDBlack) [inline include/ops/declarable/generic/broadcastable/mod.cpp]: Op not implemented? As to what it *should* be - good question...
It's obviously not contiously differentiable, but it is non-zero in most places.

File: include/ops/declarable/generic/broadcastable/reverse_mod.cpp [hunk 5]
```diff
@@ -85,5 +85,41 @@ namespace nd4j {
 
             return shapeList;
         }
+
+        CUSTOM_OP_IMPL(reversemod_bp, 3, 2, false, 0, 0) {
+            // PLEASE NOTE: we're just passing eps down the line here
```
Comment (AlexDBlack) [inline include/ops/declarable/generic/broadcastable/reverse_mod.cpp]: As per earlier comment.

File: include/ops/declarable/generic/shape/tile_to_shape.cpp [hunk 6]
```diff
@@ -0,0 +1,37 @@
+//
+//  @author raver119@gmail.com
+//
+
+#include <ops/declarable/headers/shape.h>
+
+namespace nd4j {
+namespace ops {
+    CUSTOM_OP_IMPL(tile_to_shape, 1, 1, true, 0, -1) {
```
Comment (AlexDBlack) [inline include/ops/declarable/generic/shape/tile_to_shape.cpp]: I think we'll need a tile backprop op eventually?

File: include/ops/declarable/generic/broadcastable/floordiv.cpp (hunk 2 above)
Comment (raver119) [reply include/ops/declarable/generic/broadcastable/floordiv.cpp]: yes, will remove

File: include/ops/declarable/generic/broadcastable/add.cpp (hunk 1 above)
Comment (raver119) [reply include/ops/declarable/generic/broadcastable/add.cpp]: mmm, i should check signatures. if not - we'll add it.

File: include/ops/declarable/generic/broadcastable/floormod.cpp (hunk 3 above)
Comment (raver119) [reply include/ops/declarable/generic/broadcastable/floormod.cpp]: same answer

File: include/ops/declarable/generic/shape/tile_to_shape.cpp (hunk 6 above)
Comment (raver119) [reply include/ops/declarable/generic/shape/tile_to_shape.cpp]: mmm, makes sense. will add it to this PR.

File: include/ops/declarable/generic/broadcastable/mod.cpp (hunk 4 above)
Comment (raver119) [reply include/ops/declarable/generic/broadcastable/mod.cpp]: what do you mean? impl is right there, lines 15-20. we have helpers for that.

File: include/ops/declarable/generic/broadcastable/reverse_mod.cpp (hunk 5 above)
Comment (raver119) [reply include/ops/declarable/generic/broadcastable/reverse_mod.cpp]: as above

File: include/ops/declarable/generic/broadcastable/mod.cpp (hunk 4 above)
Comment (AlexDBlack) [reply include/ops/declarable/generic/broadcastable/mod.cpp]: Ah, sorry. read that as mod_bp and was confused by the commented out code.

**Pull request merged**